### ISSUE TITLE
FastBuild affected by serious Windows10 CloseProcess performance bug

### ISSUE THREAD

**Issue by missmah** [fastbuild/fastbuild#264]
https://randomascii.wordpress.com/2017/07/09/24-core-cpu-and-i-cant-move-my-mouse/

TLDR; Windows10 has a bug where destroying processes is serialized to one CPU core with all threads attempting to close a process simultaneously fighting for the lock. The mouse is also unable to move during these lock contention storms, which can last for seconds of real time (and minutes of CPU time) on a 48-thread machine. The contention is significantly worse on higher core-count machines. It also gets worse over time, necessitating frequent (>= daily) reboots to minimize the delays.

I've experienced this bug first hand on a 72-core machine. It's quite bad and requires daily reboots.

I wonder whether it would be possible for fastbuild to simultaneously mitigate this bug in the short term, and improve performance overall, by re-using msbuild/cl.exe processes for more than one compile job on Windows?

I know VisualStudio does this internally for performance reasons, though I am unfamiliar with the details.

**Comment by ffulin** [comment]
I hope now that Bruce has reported this bug we can see a fix for it soon. I know some people who are also impacted by it, though exactly what combination of things triggers it is unclear - i.e. I run FASTBuild with Windows 10 on the same codebase as others who are affected I have not experienced any problems.

In terms of re-using cl.exe instances as a work around, I don't think that's workable for a few reasons:
 - All objects to be compiled must be specified up front, which would destroy FASTBuild's ability to load balance across CPU cores
 - Object passed on the command line must have the same compilation settings, which would complicate process launching
 - The overhead (this bug aside) of creating and destroying processes in normally negligible compared to cost of useful work, so I don't think it would be an optimization
 - I don't believe there is a way to specify multiple files for pre-processing, and similarly I don't think it's compatible with /showincludes, so dependency tracking would be problematic

**Comment by ffulin** [comment]
For anyone referencing/following this ticket, it appears this bug has been fixed and is making it's way through the Windows 10 release process. 

It's not super easy to work out if any given version of Windows 10 has this bug, but Bruce Dawson (poster of the report linked above) made a nice video: https://www.youtube.com/watch?v=cbg5O2Kbb9A

**Comment by ffulin** [comment]
It appears that the Windows 10 bug has been fixed now (https://randomascii.wordpress.com/2017/07/09/24-core-cpu-and-i-cant-move-my-mouse/)
